commit 40763baabcda86888895372cb2b14cdbbdb3d21e
Author: Trent Mick <trentm@gmail.com>
Date:   2019-06-11T14:57:38-07:00 (4 months ago)
    
    TRITON-1739 CNAPI CommandExecute still tries to use cn-agent on a factoryreset server

diff --git a/lib/workflows/server-setup.js b/lib/workflows/server-setup.js
index ca85e0a..fa2b9f7 100644
--- a/lib/workflows/server-setup.js
+++ b/lib/workflows/server-setup.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2019, Joyent, Inc.
+ * Copyright 2019 Joyent, Inc.
  */
 
 /*
@@ -24,7 +24,7 @@
  * - Set server setup => true
  */
 
-var VERSION = '1.0.4';
+var VERSION = '1.0.5';
 
 var cnapiCommon = require('wf-shared').cnapi;
 var napiCommon = require('wf-shared').napi;
@@ -75,6 +75,34 @@ function markServerAsSettingUp(job, callback) {
     });
 }
 
+/*
+ * Later tasks in this job use CNAPI CommandExecute to run setup commands on
+ * the server. CommandExecute will attempt to use cn-agent if `<server>.agents`
+ * suggests there is a sufficient cn-agent on the server. However, if the
+ * server has been `sdc-factoryreset`, then it won't have any agents, but
+ * CNAPI's server record won't know that.
+ *
+ * The `<server>.agents` will get populated later in the setup when new
+ * agents are installed.
+ */
+function clearServerAgents(job, callback) {
+    var serverUrl = '/servers/' + job.params.server_uuid;
+    var cnapi = restify.createJsonClient({ url: job.params.cnapi_url});
+
+    var update = {
+        agents: []
+    };
+    cnapi.post(serverUrl, update, function (err, req, res) {
+        if (err) {
+            job.log.info('Error clearning server.agents: ' + err.message);
+            job.log.info(err.stack.toString());
+            callback(err);
+        } else {
+            callback();
+        }
+    });
+}
+
 function fetchSetupFiles(job, callback) {
     var urUrl = '/servers/' + job.params.server_uuid + '/execute';
     var cnapiUrl = job.params.cnapi_url;
@@ -420,6 +448,12 @@ module.exports = {
             retry: 1,
             body: markServerAsSettingUp
         },
+        {
+            name: 'cnapi.clear_server_agents',
+            timeout: 10,
+            retry: 1,
+            body: clearServerAgents
+        },
         {
             name: 'cnapi.fetch_setup_files',
             timeout: 10,
diff --git a/package.json b/package.json
index 012977e..60037f2 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "cnapi",
   "description": "Triton Compute Node API",
-  "version": "1.24.1",
+  "version": "1.24.2",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
